### ISSUE TITLE
Menu item filtering

### ISSUE THREAD

**Issue by Akin909** [onivim/oni2#202]
Currently we use simplistic filtering strategy in the quick open that isn't very good at matching results or very quick, Ideally we would implement something that handles fuzzy matching a potentially huge list of results very quickly. @CrossR has been working on something like this for reason.

The strategy can be placed in `Filter.re`. If there are performance issues running this on every keystroke an alternative could be debouncing the filtering to occur only after `x` milliseconds (which I believe we did in Oni 1)

**Comment by CrossR** [comment]
I've started to have a look into this and integrating my library in now, and the first issue I've hit is that it looks like the full list isn't being passed into the sort?

That is, I've added some logic inside of `Filter.re` and the sorting wasn't changing, nor does it seem to change with the existing filter logic.

If I throw in a `Console.log(List.length(items))` in there, it seems to vary every time I open the menu, which implies that the streaming of the results could be the issue?

![image](https://user-images.githubusercontent.com/10038688/55292081-419a7f00-53de-11e9-9d20-9cc90d6ec702.png)

In the image I'm opening the menu 5 times and searching for `R` and getting different list lengths every time. It doesn't seem to change if I wait for a second or so either.

I think this would be why @bryphe was seeing the list not changing since only part of it is being updated.

**Comment by Akin909** [comment]
@CrossR each time you hit add a letter in the quick open you will be filtering against a potentially ongoing streaming in of results

![ripgrep_results](https://user-images.githubusercontent.com/22454918/55292317-1d8c6d00-53e1-11e9-9491-cf6ac3f4b1a1.png)

The screenshot shows the incremental accumulation of results, regarding the results not changing, I haven't been able to reproduce this locally but the current mechanism is very fussy i.e. it expects an exact match i.e. if you want to match on `Filter.re` you have to type that string in exactly it cant be `Fltr` that will get no results.

**Comment by Akin909** [comment]
As you say I wouldn't expect a consistent number when re-opening the list as it could be some arbitrary amount being streamed in at a time.

The filter is being passed a `command` type which is a record
```
type command = {
 name: string,
 command: unit => unit,
 icons: option(string)
}
```
so you have to filter on `name`.

**Comment by Akin909** [comment]
One thing to note @CrossR is that the sequence of events is
* type a key -> `MenuQuery(userQueryString)` -> `Filter.menu(queryAtTheTime, CommandsAtTheTime)`

* Meanwhile open quickOpen -> stream begins -> intermittently calls -> `MenuUpdate(newCommands)`

the above flows are separate i.e. the filter is *currently* only called when a user hits a key not when the content of the list changes (I didn't forsee us using a streaming pattern), it might be that in the reducer in `Menu.re` when an update comes in you should also check if `String.length(query) > 0` if so append the commands but also, re-run `Filter.menu(existingQuery, newlyUpdatedCommands)`

**Comment by CrossR** [comment]
I think we can close this out, and make any follow up issues if we hit any performance issues or scoring issues.

Implemented over in #204.
